### PR TITLE
(dsl): Support `regexp` query

### PR DESCRIPTION
**Pull request by milicns**
Part of #91

### CODE REVIEW

File: modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala [hunk 1]
```diff
@@ -1050,6 +1050,62 @@ object HttpExecutorSpec extends IntegrationSpec {
...
+                  query =
+                    ElasticQuery.regexp(
+                      field = TestDocument.stringField,
+                      value =
+                        s"${firstDocument.stringField.take(1)}.*${firstDocument.stringField.takeRight(1)}".toUpperCase
+                    )
+                  res <- Executor
+                           .execute(ElasticRequest.search(firstSearchIndex, query))
+                           .documentAs[TestDocument]
+                } yield assert(res)(!Assertion.contains(firstDocument))
```
Comment (dbulaja98) [inline modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala:1075]: It does not make sense to me. Should it maybe be `assert(res)(Assertion.contains(firstDocument))`?

Comment (dbulaja98) [reply modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala:1075]: It probably fails.

File: modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala [hunk 2]
```diff
@@ -1050,6 +1050,62 @@ object HttpExecutorSpec extends IntegrationSpec {
...
+                            .caseInsensitiveTrue
+                  res <- Executor
+                           .execute(ElasticRequest.search(firstSearchIndex, query))
+                           .documentAs[TestDocument]
+                } yield (assert(res)(Assertion.contains(firstDocument)) && assert(res)(!Assertion.contains(secondDocument)))
+            }
+          } @@ around(
+            Executor.execute(ElasticRequest.createIndex(firstSearchIndex)),
+            Executor.execute(ElasticRequest.deleteIndex(firstSearchIndex)).orDie
+          ),
```
Comment (dbulaja98) [inline modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala:1108]: I think you can remove this test.

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 3]
```diff
@@ -621,6 +621,36 @@ object ElasticQuery {
   final def range(field: String): RangeQuery[Any, Any, Unbounded.type, Unbounded.type] =
     Range.empty[Any, Any](field = field)
 
+  /**
+   * Constructs a type-safe instance of [[zio.elasticsearch.query.RegexpQuery]] using the specified parameters.
+   * [[zio.elasticsearch.query.RegexpQuery]] returns documents that contain terms matching a regular expression.
+   *
+   * @param field
+   *   the field for which query is specified for
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:629]:
```suggestion
   *   the type-safe field for which query is specified for
```

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 4]
```diff
@@ -1165,6 +1165,17 @@ object ElasticQuerySpec extends ZIOSpecDefault {
             )
           )
         },
+        test("regexp") {
+          val query                      = regexp("stringField", "t.*st")
+          val queryTs                    = regexp(TestDocument.stringField, "t.*st")
+          val queryWithCaseInsensitive   = regexp("stringField", "t.*st").caseInsensitiveTrue
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:1171]: You don't need this one.

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 5]
```diff
@@ -1165,6 +1165,17 @@ object ElasticQuerySpec extends ZIOSpecDefault {
...
+          val queryTsWithCaseInsensitive = regexp(TestDocument.stringField, "t.*st").caseInsensitiveTrue
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:1172]: Rename this one to `queryWithCaseInsensitive`.

Comment (dbulaja98) [reply modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:1172]: And add another one val `queryWithSuffix`, where you can do something like this: `TestDocument.stringField.raw`.

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 6]
```diff
@@ -2772,6 +2783,40 @@ object ElasticQuerySpec extends ZIOSpecDefault {
           assert(queryMixedBoundsWithBoost.toJson(fieldPath = None))(equalTo(expectedMixedBoundsWithBoost.toJson)) &&
           assert(queryWithFormat.toJson(fieldPath = None))(equalTo(expectedWithFormat.toJson))
         },
+        test("regexp") {
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:2786]: Do the same as above here.

**Review by dbulaja98 — CHANGES_REQUESTED** [review]

File: modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala (hunk 1 above)
Comment (milicns) [reply modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala:1075]: We agreed to just change the name of the test.

File: modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala (hunk 2 above)
Comment (milicns) [reply modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala:1108]: Agreed to just change the name of the previous test test.

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 7]
```diff
@@ -1165,6 +1165,23 @@ object ElasticQuerySpec extends ZIOSpecDefault {
...
+          assert(query)(equalTo(Regexp[Any](field = "stringField", value = "t.*st", caseInsensitive = None))) &&
+          assert(queryTs)(
+            equalTo(Regexp[TestDocument](field = "stringField", value = "t.*st", caseInsensitive = None))
+          ) &&
+          assert(queryWithCaseInsensitive)(
+            equalTo(Regexp[TestDocument](field = "stringField", value = "t.*st", caseInsensitive = Some(true)))
+          ) &&
+          assert(queryWithSuffix)(
+            equalTo(Regexp[TestDocument](field = "stringField.raw", value = "t.*st", caseInsensitive = None))
+          )
```
Comment (milicns) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:1183]: Sbt prepare divide those asserts in the new lines

**Review by dbulaja98 — APPROVED** [review]

**Pull request merged**